network/netidentity.go

```diff
@@ -0,0 +1,412 @@
+	signedChal := i.Sign(s)
+	return base64.StdEncoding.EncodeToString(protocol.Encode(&signedChal))
+}
+
+func (i identityChallenge) Sign(secrets *crypto.SignatureSecrets) identityChallengeSigned {
+	return identityChallengeSigned{Msg: i, Signature: secrets.Sign(i)}
+}
+
+func (i identityChallenge) ToBeHashed() (protocol.HashID, []byte) {
+	return protocol.NetIdentityChallenge, protocol.Encode(&i)
+}
+
+// Verify checks that the signature included in the identityChallenge was indeed created by the included Key
+func (i identityChallengeSigned) Verify() bool {
+	return i.Msg.Key.Verify(i.Msg, i.Signature)
+}
+
+func (i identityChallengeResponse) signAndEncodeB64(s *crypto.SignatureSecrets) string {
+	signedChalResp := i.Sign(s)
+	return base64.StdEncoding.EncodeToString(protocol.Encode(&signedChalResp))
+}
+
+func (i identityChallengeResponse) Sign(secrets *crypto.SignatureSecrets) identityChallengeResponseSigned {
+	return identityChallengeResponseSigned{Msg: i, Signature: secrets.Sign(i)}
+}
+
+func (i identityChallengeResponse) ToBeHashed() (protocol.HashID, []byte) {
+	return protocol.NetIdentityChallengeResponse, protocol.Encode(&i)
+}
+
+// Verify checks that the signature included in the identityChallengeResponse was indeed created by the included Key
+func (i identityChallengeResponseSigned) Verify() bool {
+	return i.Msg.Key.Verify(i.Msg, i.Signature)
+}
+
+func (i identityVerificationMessage) Sign(secrets *crypto.SignatureSecrets) identityVerificationMessageSigned {
+	return identityVerificationMessageSigned{Msg: i, Signature: secrets.Sign(i)}
+}
+
+func (i identityVerificationMessage) ToBeHashed() (protocol.HashID, []byte) {
+	return protocol.NetIdentityVerificationMessage, protocol.Encode(&i)
+}
+
+// Verify checks that the signature included in the identityVerificationMessage was indeed created by the included Key
+func (i identityVerificationMessageSigned) Verify(key crypto.PublicKey) bool {
+	return key.Verify(i.Msg, i.Signature)
+}
+
+// identityVerificationHandler receives a signature over websocket, and confirms it matches the
+// sender's claimed identity and the challenge that was assigned to it. If the identity is available,
+// the peer is loaded into the identity tracker. Otherwise, we ask the network to disconnect the peer.
+func identityVerificationHandler(message IncomingMessage) OutgoingMessage {
+	peer := message.Sender.(*wsPeer)
+	// avoid doing work (crypto and potentially taking a lock) if the peer is already verified
+	if atomic.LoadUint32(&peer.identityVerified) == 1 {
+		return OutgoingMessage{}
+	}
+	localAddr, _ := peer.net.Address()
+	msg := identityVerificationMessageSigned{}
+	err := protocol.Decode(message.Data, &msg)
+	if err != nil {
+		networkPeerIdentityError.Inc(nil)
+		peer.net.log.With("err", err).With("remote", peer.OriginAddress()).With("local", localAddr).Warn("peer identity verification could not be decoded, disconnecting")
+		peer.net.disconnect(peer, disconnectBadIdentityData)
```

Comment:
why disconnecting right here instead of returning Action=Disconnect as other handler do?